_posts/2017-03-28-comparison-of-regex-engines.md

```diff
@@ -0,0 +1,226 @@
+provides functions for parsing, compiling, and executing regular expressions:
+> Its syntax is similar to Perl-style regular expressions, but lacks a few features like look around and back-references.
+In exchange, all searches execute in linear time with respect to the size of the regular expression and search text.
+
+Except the Rust crate all engines are written in C or C++ including the test tool. It was a requirement to have
+C-binding of the used engine. So an interface was necessary to call the Rust functions within C.
+The solution uses the Rusts FFI (foreign function interface) to build a static
+library which just counts the matches for a given expression. The complete library consists of 3 functions with less
+than 50 lines of code. The main Rust function to get the matches is:
+```rust
+#[no_mangle]
+pub extern fn regex_matches(raw_exp: *mut Regex, p: *const u8, len: u64) -> u64 {
+    let exp = unsafe { Box::from_raw(raw_exp) };
+    let s = unsafe { slice::from_raw_parts(p, len as usize) };
+
+    let findings = exp.find_iter(s).count();
+    Box::into_raw(exp);
+    findings as u64
+}
+```
+The function gets a raw pointer to a previously compiled expression (`raw_exp`), a raw pointer to the input c-string
+(`p`) and the length of the input string (`len`). At the first the function gets the compiled expression from the
+corresponding raw pointer and the input string. The converting from a raw pointer into a type is unsafe, that's why
+the code parts have to be wrapped with `unsafe {}`. After all the number of matches is gathered with a call of
+`exp.find_iter(s).count()`. To use the compiled expression in following function calls, the raw
+pointer from the expression is gathered again. This effects that the lifetime of the expression still remains after returning.
+At the end the function returns the number of matches as a 64bit value to the caller.
+
+The corresping C function prototype is:
+```c
+struct Regex;       // anonymous declaration
+
+extern uint64_t regex_matches(struct Regex const * const exp, uint8_t * const str, uint64_t str_len);
+```
+
+# Results
+To get the test results of the test tool call within the build directory of the tool:
+```bash
+./src/regex_perf -f ../3200.txt -o results.csv
+```
+The tool prints the details and writes the results per test and engine to `results.csv`.
+At the end a short summary of the results is printed:
+```
+Total Results:
+[      pcre] time:  12626.7 ms, score:      8 points,
+[  pcre-dfa] time:  14135.2 ms, score:      0 points,
+[  pcre-jit] time:   1050.6 ms, score:     47 points,
+[       re2] time:    946.1 ms, score:     26 points,
+[      onig] time:   2475.8 ms, score:      4 points,
+[       tre] time:  10508.4 ms, score:      0 points,
+[     hscan] time:    299.7 ms, score:     72 points,
+[rust_regex] time:   3681.5 ms, score:     47 points,
+```
+
+## Timings
+On the basis of a CSV-file I did some analytics. At first I summarized the overall execution time per engine.
```

Comment:
I would keep the present here and not switch to past. Also for further sentences. :)